### PR TITLE
Fix issue #108

### PR DESCRIPTION
**Pull request by geniusgordon**

### CODE REVIEW

File: problem/templates/problem/problem_list.html [hunk 1]
```diff
@@ -34,32 +34,48 @@
...
-          <span id='rejudge' class='glyphicon glyphicon-refresh' 
-          data-toggle="tooltip" data-placement="left" title="Rejudge"></span>&nbsp;{{p.pid}}</td>
-        <td><a href="{% url 'problem:problem.views.detail' p.pid %}">{{p.name}}</a>&nbsp;
-          <a href="{% url 'problem:problem.views.edit' p.pid %}">
-            <span id='edit' class='glyphicon glyphicon-pencil' 
-            data-toggle="tooltip" data-placement="right" title="Edit"></span>
-          </a>
+          {% if user == p.owner or user.is_admin %}
+            <span id='rejudge' class='glyphicon glyphicon-refresh' 
+            data-toggle="tooltip" data-placement="left" title="Rejudge"></span>
```
Comment (nickelcheng) [inline problem/templates/problem/problem_list.html]: I prefer ```data-toggle``` aligns with ```id```

File: problem/templates/problem/problem_list.html [hunk 2]
```diff
@@ -34,32 +34,48 @@
...
-              {{p.not_pass}}%
+            <div class="progress-bar progress-bar-danger" style="width: {{p.not_pass_rate}}%;">
+              {{p.not_pass_rate}}%
             </div>
           </div>
         </td>
         <td>
-          <button class="btn btn-primary">Delete</button>
+          {% if user == p.owner or user.is_admin %}
+            <a data-target="{{p.pk}}" class="btn btn-primary" href="{% url 'problem:delete_problem' p.pk %}">Delete</a>
```
Comment (nickelcheng) [inline problem/templates/problem/problem_list.html]: This line is too long.

**Comment by nickelcheng** [comment]
1. Preview tag and actual tag is still not match.
2. After choosing a judge type, I can not modify it anymore (the list becomes empty)
3. We need a judge language choice radix button
4. Live archive and POJ can not fill in ```other judge id``` QQ
5. New problem name is suggested to filled in problem panel page, put it beside the add problem button.

**Comment by nickelcheng** [comment]
Another thing: in case of add / edit / delete problem or testcase, please record the executor in our log.
Ex: ```User henry add problem 10000```

**Comment by geniusgordon** [comment]
fixed

File: problem/forms.py [hunk 3]
```diff
@@ -24,13 +24,14 @@
...
 from users.models import User
+from django.db.models import Q
 
 import autocomplete_light
 
 # create autocomplete interface and register
 class UserAutocomplete(autocomplete_light.AutocompleteModelBase):
     search_fields = ['^username']
-    choices = User.objects.all()
+    choices = User.objects.filter(Q(user_level=User.ADMIN) | Q(user_level=User.JUDGE) | Q(user_level=User.SUB_JUDGE))
```
Comment (nickelcheng) [inline problem/forms.py]: This line is too long.

File: problem/static/problem/js/edit.js [hunk 4]
```diff
@@ -121,31 +122,74 @@ function add_new_testcase(pid, data) {
...
-    if (this.value == "ERR_TORRENT")
-        $("#id_error_torrence").parent().show();
-    else if (this.value == "OTHER")
-        $("#id_other_judge_id").parent().show();
-    else if (this.value == "PARTIAL")
-        $("#id_partial_judge_code").parent().show();
-    else if (this.value == "SPECIAL")
-        $("#id_special_judge_code").parent().show();
+    if (option == "ERR_TOLERANT")
+        $("#id_error_torrence").parent().parent().show();
```
Comment (nickelcheng) [inline problem/static/problem/js/edit.js]: typo: torrence -> tolerance

File: problem/static/problem/js/edit.js [hunk 5]
```diff
@@ -121,31 +122,74 @@ function add_new_testcase(pid, data) {
...
+    form.attr("action", "");
+    return true;
 });
 
 function hide_field() {
-  $("#id_error_torrence").parent().hide();
-  $("#id_other_judge_id").parent().hide();
-  $("#id_partial_judge_code").parent().hide();
-  $("#id_special_judge_code").parent().hide();
+  $("#id_error_torrence").parent().parent().hide();
```
Comment (nickelcheng) [inline problem/static/problem/js/edit.js]: typo: torrence -> tolerance

**Comment by nickelcheng** [comment]
1. Other judge id should not be negative.
2. Judge Language field should only appear when judge type = special judge or partial judge
3. I can not add tags in preview page, is it correct? If so, please remove add tag button in preview page.
4. I prefer to use auto-complete for add tag field.
5. Problem status not ready yet?
---
Not fixed comments
* New problem name is suggested to filled in problem panel page, put it beside the add problem button.
* In case of add / edit / delete problem or testcase, please record the executor in our log.

**Comment by geniusgordon** [comment]
fixed

**Comment by nickelcheng** [comment]
I got NoReverseMatch
error

**Comment by geniusgordon** [comment]
fixed

**Comment by geniusgordon** [comment]
add submit

**Comment by nickelcheng** [comment]
1. In case of add / edit / delete problem or testcase, please record the executor in our log.
2. Preview button GG
error

3. Add problem log disappears!! Only edit problem log left.

**Comment by geniusgordon** [comment]
fixed

**Comment by geniusgordon** [comment]
add problem, tag log

File: problem/views.py [hunk 6]
```diff
@@ -39,62 +41,75 @@
...
+    return render_index(request, 'problem/detail.html', {'problem': problem, 'tag_form': tag_form})
+
+@login_required
+def new(request):
+    if request.method == "GET":
+        return render_index(request, "problem/new.html")
+    elif request.method == "POST":
+        if 'pname' in request.POST and request.POST['pname'].strip() != "":
+            p = Problem(pname=request.POST['pname'], owner=request.user)
+            p.save()
```
Comment (nickelcheng) [inline problem/views.py]: Please add info log.

File: problem/views.py [hunk 7]
```diff
@@ -103,35 +118,31 @@ def edit(request, pid=None):
...
             problem = Problem.objects.get(pk=pid)
         except Problem.DoesNotExist:
             logger.warning("problem %s does not exist" % (pid))
             raise Http404("problem %s does not exist" % (pid))
         if not problem.tags.filter(tag_name=tag).exists():
-            logger.info("add new tag '%s' to problem %s" % (tag, pid))
             new_tag, created = Tag.objects.get_or_create(tag_name=tag)
             problem.tags.add(new_tag)
             problem.save()
+            logger.info("add new tag '%s' to problem %s by" % (tag, pid, request.user))
```
Comment (nickelcheng) [inline problem/views.py]: ```"add new tag '%s' to problem %s by %s"```

**Comment by geniusgordon** [comment]
fixed

**Comment by nickelcheng** [comment]
+1

**Comment by nickelcheng** [comment]
Please merge the latest code.

**Comment by geniusgordon** [comment]
merged latest code

**Comment by nickelcheng** [comment]
going to merge.

**Pull request merged**